protocols/mempool-overview.md

```diff
@@ -134,10 +139,134 @@ Committed transactions are not in the mempool:
+_(ii)_ invalid transactions are _not_ kept in the cache, that is the parameter `KeepInvalidTxsInCache` is always set to false, and
+_(iii)_ the `flush` operation is never called.
+In addition, we shall assume that a finite amount of transactions is received by the system, and that if a transaction is received at a correct process, then it is eventually received at all the correct processes (thanks to the gossip layer in `mempool/reactor.go`).
+
+Then, let us consider some transaction $tx$.
+According to the logic in `clist_mempool.go`,
+
+$tx$ is added to the cache at time $t$ (in short, @t) if  
+(1) $tx$ is received and not already there @t (`checkTx`, [l237](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L237)), or  
+(2) $tx$ is committed and valid @$t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L598)).
+
+$tx$ is removed from the cache @t if  
+(3) $tx$ was received @t'<t, re-checked and invalid t'<@t''<t (`resCbRecheck`, [l477](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L4747)), or  
+(4) $tx$ is ejected from the cache as it is the last transaction wrt. cache and a new transaction is added, or  
+(5) $tx$ is committed and invalid @t'<t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L601)).
+
+$tx$ is added to the mempool @t if  
+(6) $tx$ was received and added to the cache @t'<t, and was valid t'<@t"<t  (`addTx` after `recvCbFirstTime`, [l318](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L318)).
+
+$tx$ is removed from the mempool @t if  
+(7) $tx$ is committed @t (`RemoveTx` after `removeTxByKey` after `update`, [l614](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L614)), or  
+(8) $tx$ is re-checked and invalid @t'<t (`removeTx` after `resCbRecheck`, [l474](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L474)).
+
+In addition, following the logic of CometBFT,  
+- check and re-check operations are asynchronous,
+- before committing transactions, checks and re-checks are frozen and all the pending calls return, and
+- upon commit, all the transactions remaining in the mempool are re-checked.
+
+**Correctness**
+We now establish that the above algorithm implements the mempool abstraction.
+For invariants INV2a and INV3a, we show that **there exist** such an $\alpha$ and $\beta$, but we do not look at characterizing them precisely.
+Regarding INV4, we shall assume that step (1) occurs infinitely often and not once upon the reception of the transaction.
+This models that a CometBFT client may re-submit a failed transaction forever until it succeeds (if it does).
```

Comment:
What is a "failed" transaction here? I didn't fully understand the statement before, referring to (1).